_typesInDynamicallyAccessedMembersHierarchy might get modified from a readonly enumeration

There is a CI test failure that indicates a dictionary getting modified while enumerating via a foreach loop. The failure repro's locally when running all the tests but I haven't been able to narrow it down to running a single failing test.

Given that the code does modify _typesInDynamicallyAccessedMembersHierarchy at ApplyDynamicallyAccessedMembersToTypeHierarchyInner() while enumerating, I think it might be faster to change the code to not do that.

**Command to run the full tests (should see some tests failing):** 

1. Sync to the PR, https://github.com/dotnet/runtime/pull/51237,  and build locally

2. build.cmd -ci -arch x64 -os windows -s libs.tests -c Release /p:TestAssemblies=false /p:TestTrimming=true  /p:CrossBuild=false /bl:TrimmingTests.binlog

Should be able to see test, EventSourcePropertyValueTest.cs, failing among others

**Steps that I tried to repro a single test failure, EventSourcePropertyValueTest (have not been able to repro)**

1. Open the TrimmingTests.binlog, search for illink failure related to EventSourcePropertyValueTest
2. Copy the illink failure commandline and run

**Failure Callstack**

##[error]ILLink(0,0): error IL1005: (NETCORE_ENGINEERING_TELEMETRY=Build) System.Diagnostics.Tracing.EventSource.<EnsureDescriptorsInitialized>g__GetCreateManifestAndDescriptorsViaLocalMethod|93_0(String): Error processing method 'System.Diagnostics.Tracing.EventSource.<EnsureDescriptorsInitialized>g__GetCreateManifestAndDescriptorsViaLocalMethod|93_0(String)' in assembly 'System.Private.CoreLib.dll'
  Mono.Linker.LinkerFatalErrorException: ILLink: error IL1005: System.Diagnostics.Tracing.EventSource.<EnsureDescriptorsInitialized>g__GetCreateManifestAndDescriptorsViaLocalMethod|93_0(String): Error processing method 'System.Diagnostics.Tracing.EventSource.<EnsureDescriptorsInitialized>g__GetCreateManifestAndDescriptorsViaLocalMethod|93_0(String)' in assembly 'System.Private.CoreLib.dll'
   ---> System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
     at System.Collections.Generic.Dictionary`2.Enumerator.MoveNext()
     at Mono.Linker.Dataflow.DynamicallyAccessedMembersTypeHierarchy.ApplyDynamicallyAccessedMembersToTypeHierarchy(ReflectionMethodBodyScanner reflectionMethodBodyScanner, ReflectionPatternContext& reflectionPatternContext, TypeDefinition type)
     at Mono.Linker.Dataflow.ReflectionMethodBodyScanner.HandleCall(MethodBody callingMethodBody, MethodReference calledMethod, Instruction operation, ValueNodeList methodParams, ValueNode& methodReturnValue)
     at Mono.Linker.Dataflow.MethodBodyScanner.HandleCall(MethodBody callingMethodBody, Instruction operation, Stack`1 currentStack)
     at Mono.Linker.Dataflow.MethodBodyScanner.Scan(MethodBody methodBody)
     at Mono.Linker.Dataflow.ReflectionMethodBodyScanner.ScanAndProcessReturnValue(MethodBody methodBody)
     at Mono.Linker.Steps.MarkStep.MarkMethodBody(MethodBody body)
     at Mono.Linker.Steps.MarkStep.ProcessMethod(MethodDefinition method, DependencyInfo& reason)
     at Mono.Linker.Steps.MarkStep.ProcessQueue()
     --- End of inner exception stack trace ---
     at Mono.Linker.Steps.MarkStep.ProcessQueue()
     at Mono.Linker.Steps.MarkStep.ProcessPrimaryQueue()
     at Mono.Linker.Steps.MarkStep.Process()
     at Mono.Linker.Steps.MarkStep.Process(LinkContext context)
     at Mono.Linker.Pipeline.ProcessStep(LinkContext context, IStep step)
     at Mono.Linker.Pipeline.Process(LinkContext context)
     at Mono.Linker.Driver.Run(ILogger customLogger)
  Optimizing assemblies for size, which may change the behavior of the app. Be sure to test after publishing. See: https://aka.ms/dotnet-illink
D:\workspace\_work\1\s\.dotnet\sdk\6.0.100-preview.2.21155.3\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.ILLink.targets(59,5): error NETSDK1144: Optimizing assemblies for size failed. Optimization can be disabled by setting the PublishTrimmed property to false. [D:\workspace\_work\1\s\artifacts\bin\trimmingTests\projects\Microsoft.Extensions.Http.TrimmingTests\AddTypedClientTests\win-x64\project.csproj]
##[error].dotnet\sdk\6.0.100-preview.2.21155.3\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.ILLink.targets(59,5): error

--- a/src/linker/Linker.Dataflow/DynamicallyAccessedMembersTypeHierarchy.cs
+++ b/src/linker/Linker.Dataflow/DynamicallyAccessedMembersTypeHierarchy.cs
@@ -89,15 +89,6 @@ namespace Mono.Linker.Dataflow
 
 			Debug.Assert (!apply || annotation != DynamicallyAccessedMemberTypes.None);
 
-			if (apply) {
-				// One of the base/interface types is already marked as having the annotation applied
-				// so we need to apply the annotation to this type as well
-				var reflectionMethodBodyScanner = new ReflectionMethodBodyScanner (_context, _markStep);
-				var reflectionPatternContext = new ReflectionPatternContext (_context, true, type, type);
-				reflectionMethodBodyScanner.ApplyDynamicallyAccessedMembersToType (ref reflectionPatternContext, type, annotation);
-				reflectionPatternContext.Dispose ();
-			}
-
 			// Store the results in the cache
 			// Don't store empty annotations for non-interface types - we can use the presence of the row
 			// in the cache as indication of it instead.
@@ -105,6 +96,20 @@ namespace Mono.Linker.Dataflow
 			// already filled), so we need to always store the row (even if empty) for interfaces.
 			if (annotation != DynamicallyAccessedMemberTypes.None || type.IsInterface) {
 				_typesInDynamicallyAccessedMembersHierarchy[type] = (annotation, apply);
+			}
+
+			// It's important to first store the annotation in the cache and only then apply the annotation.
+			// Applying the annotation will lead to marking additional types which in turn calls back into this
+			// method to look for annotations. If the newly marked type derives from the one we're processing
+			// it will rely on the cache to know if it's annotated - so the record must be in the cache
+			// before it happens.
+			if (apply) {
+				// One of the base/interface types is already marked as having the annotation applied
+				// so we need to apply the annotation to this type as well
+				var reflectionMethodBodyScanner = new ReflectionMethodBodyScanner (_context, _markStep);
+				var reflectionPatternContext = new ReflectionPatternContext (_context, true, type, type);
+				reflectionMethodBodyScanner.ApplyDynamicallyAccessedMembersToType (ref reflectionPatternContext, type, annotation);
+				reflectionPatternContext.Dispose ();
 			}
 
 			return (annotation, apply);
@@ -134,11 +139,32 @@ namespace Mono.Linker.Dataflow
 			// Propagate the newly applied annotation to all derived/implementation types
 			// Since we don't have a data structure which would allow us to enumerate all derived/implementation types
 			// walk all of the types in the cache. These are good candidates as types not in the cache don't apply.
-			foreach (var candidate in _typesInDynamicallyAccessedMembersHierarchy) {
-				if (candidate.Value.annotation == DynamicallyAccessedMemberTypes.None)
-					continue;
+			//
+			// Applying annotations can lead to marking additional types which can lead to adding new records
+			// to the cache. So we can't simply iterate over the cache. We also can't rely on the auto-applying annotations
+			// which is triggered from marking via ProcessMarkedTypeForDynamicallyAccessedMembersHierarchy as that will
+			// only reliably work once the annotations are applied to all types in the cache first. Partially
+			// applied annotations to the cache are not enough. So we have to apply the annotations to any types
+			// added to the cache during the application as well.
+			//
+			HashSet<TypeDefinition> typesProcessed = new HashSet<TypeDefinition> ();
+			List<TypeDefinition> candidateTypes = new List<TypeDefinition> ();
+			while (true) {
+				candidateTypes.Clear ();
+				foreach (var candidate in _typesInDynamicallyAccessedMembersHierarchy) {
+					if (candidate.Value.annotation == DynamicallyAccessedMemberTypes.None || candidate.Value.applied)
+						continue;
 
-				ApplyDynamicallyAccessedMembersToTypeHierarchyInner (reflectionMethodBodyScanner, ref reflectionPatternContext, candidate.Key);
+					if (typesProcessed.Add (candidate.Key))
+						candidateTypes.Add (candidate.Key);
+				}
+
+				if (candidateTypes.Count == 0)
+					break;
+
+				foreach (var candidateType in candidateTypes) {
+					ApplyDynamicallyAccessedMembersToTypeHierarchyInner (reflectionMethodBodyScanner, ref reflectionPatternContext, candidateType);
+				}
 			}
 
 			return annotation;

--- a/test/Mono.Linker.Tests.Cases/Reflection/ObjectGetType.cs
+++ b/test/Mono.Linker.Tests.Cases/Reflection/ObjectGetType.cs
@@ -7,6 +7,7 @@ using System.Collections.Generic;
 using System.Diagnostics.CodeAnalysis;
 using System.Text;
 using Mono.Linker.Tests.Cases.Expectations.Assertions;
+using Mono.Linker.Tests.Cases.Expectations.Helpers;
 using Mono.Linker.Tests.Cases.Expectations.Metadata;
 
 namespace Mono.Linker.Tests.Cases.Reflection
@@ -43,6 +44,11 @@ namespace Mono.Linker.Tests.Cases.Reflection
 
 			DiamondShapeWithUnannotatedInterface.Test ();
 			DiamondShapeWithAnnotatedInterface.Test ();
+
+			ApplyingAnnotationIntroducesTypesToApplyAnnotationTo.Test ();
+			ApplyingAnnotationIntroducesTypesToApplyAnnotationToViaInterfaces.Test ();
+			ApplyingAnnotationIntroducesTypesToApplyAnnotationToMultipleAnnotations.Test ();
+			ApplyingAnnotationIntroducesTypesToApplyAnnotationToEntireType.Test ();
 		}
 
 		[Kept]
@@ -942,6 +948,281 @@ namespace Mono.Linker.Tests.Cases.Reflection
 			public static void Test ()
 			{
 				GetInstance ().GetType ().GetMethod ("InterfaceMethod");
+			}
+		}
+
+		[Kept]
+		class ApplyingAnnotationIntroducesTypesToApplyAnnotationTo
+		{
+			[Kept]
+			[KeptMember (".ctor()")]
+			[KeptAttributeAttribute (typeof (DynamicallyAccessedMembersAttribute))]
+			[DynamicallyAccessedMembers (DynamicallyAccessedMemberTypes.PublicNestedTypes)]
+			class AnnotatedBase
+			{
+			}
+
+			[Kept]
+			interface IUnannotatedInterface
+			{
+			}
+
+			interface IUnusedInterface
+			{
+			}
+
+			[Kept]
+			[KeptBaseType (typeof (AnnotatedBase))]
+			[KeptMember (".ctor()")]
+			class Derived : AnnotatedBase
+			{
+				[Kept]
+				[KeptBaseType (typeof (AnnotatedBase))]
+				public class NestedDerived : AnnotatedBase
+				{
+					[Kept]
+					[KeptBaseType (typeof (NestedDerived))]
+					public class DeepNestedDerived : NestedDerived
+					{
+						[Kept] // Marked due to the annotation
+						public class DeepNestedChild
+						{
+						}
+
+						private class DeepNestedPrivateChild
+						{
+						}
+					}
+
+					[Kept] // Marked due to the annotation
+					[KeptInterface (typeof (IUnannotatedInterface))]
+					public class NestedChild : IUnannotatedInterface
+					{
+					}
+				}
+
+				// Not used - not marked
+				private class PrivateNested : IUnusedInterface
+				{
+				}
+			}
+
+			[Kept]
+			static AnnotatedBase GetInstance () => new Derived ();
+
+			[Kept]
+			public static void Test ()
+			{
+				var t = typeof (IUnannotatedInterface);
+				GetInstance ().GetType ().GetNestedTypes ();
+			}
+		}
+
+		[Kept]
+		class ApplyingAnnotationIntroducesTypesToApplyAnnotationToViaInterfaces
+		{
+			[Kept]
+			[KeptAttributeAttribute (typeof (DynamicallyAccessedMembersAttribute))]
+			[DynamicallyAccessedMembers (DynamicallyAccessedMemberTypes.PublicFields)]
+			interface IAnnotatedInterface
+			{
+			}
+
+			[Kept]
+			[KeptMember (".ctor()")]
+			[KeptInterface (typeof (IAnnotatedInterface))]
+			class ImplementsInterface : IAnnotatedInterface
+			{
+				[Kept]
+				public FieldTypeAlsoImplementsInterface _publicFieldWithInterface;
+
+				UnusedFieldTypeImplementsInterface _privateFieldWithInterface;
+			}
+
+			[Kept]
+			[KeptMember (".ctor()")]
+			[KeptInterface (typeof (IAnnotatedInterface))]
+			class FieldTypeAlsoImplementsInterface : IAnnotatedInterface
+			{
+				[Kept]
+				public FieldTypeAlsoImplementsInterface _selfReference;
+
+				[Kept]
+				public NestedFieldType _nestedField;
+
+				[Kept]
+				public class NestedFieldType
+				{
+				}
+			}
+
+			class UnusedFieldTypeImplementsInterface : IAnnotatedInterface
+			{
+			}
+
+			[Kept]
+			static IAnnotatedInterface GetInstance () => new ImplementsInterface ();
+
+			[Kept]
+			public static void Test ()
+			{
+				var t = new FieldTypeAlsoImplementsInterface (); // Instantiate the type so that it gets interfaces
+				GetInstance ().GetType ().GetFields ();
+			}
+		}
+
+		[Kept]
+		class ApplyingAnnotationIntroducesTypesToApplyAnnotationToMultipleAnnotations
+		{
+			[Kept]
+			[KeptMember (".ctor()")]
+			[KeptAttributeAttribute (typeof (DynamicallyAccessedMembersAttribute))]
+			[DynamicallyAccessedMembers (DynamicallyAccessedMemberTypes.PublicMethods)]
+			class MethodAnnotatedBase
+			{
+			}
+
+			[Kept]
+			[KeptMember (".ctor()")]
+			[KeptBaseType (typeof (MethodAnnotatedBase))]
+			class DerivedFromMethodsBase : MethodAnnotatedBase
+			{
+				[Kept]
+				public AnotherMethodsDerived PublicMethod () { return null; }
+
+				void PrivateMethod () { }
+			}
+
+			[Kept]
+			[KeptBaseType (typeof (MethodAnnotatedBase))]
+			class AnotherMethodsDerived : MethodAnnotatedBase
+			{
+				[Kept]
+				public static void PublicStaticMethod (DerivedFromPropertiesBase p) { }
+
+				static void PrivateStaticMethod () { }
+			}
+
+			[Kept]
+			[KeptMember (".ctor()")]
+			[KeptAttributeAttribute (typeof (DynamicallyAccessedMembersAttribute))]
+			[DynamicallyAccessedMembers (DynamicallyAccessedMemberTypes.PublicProperties)]
+			class PropertiesAnnotatedBase
+			{
+			}
+
+			[Kept]
+			[KeptBaseType (typeof (PropertiesAnnotatedBase))]
+			class DerivedFromPropertiesBase : PropertiesAnnotatedBase
+			{
+				[Kept]
+				public static AnotherPropertiesDerived PublicProperty { [Kept] get => null; }
+
+				private static UnusedType PrivateProperty { get => null; }
+			}
+
+			[Kept]
+			[KeptBaseType (typeof (PropertiesAnnotatedBase))]
+			class AnotherPropertiesDerived : PropertiesAnnotatedBase
+			{
+				[Kept]
+				public static UsedType PublicProperty { [Kept] get => null; }
+
+				private static UnusedType PrivateProperty { get => null; }
+			}
+
+			[Kept]
+			class UsedType { }
+
+			class UnusedType { }
+
+			[Kept]
+			static MethodAnnotatedBase GetMethodsInstance () => new DerivedFromMethodsBase ();
+
+			[Kept]
+			static PropertiesAnnotatedBase GetPropertiesInstance () => new PropertiesAnnotatedBase ();
+
+			[Kept]
+			public static void Test ()
+			{
+				GetMethodsInstance ().GetType ().GetMethods ();
+
+				// Note that the DerivedFromPropertiesBase type is not referenced any other way then through
+				// the PublicStaticMethod parameter which is only kept due to the hierarchy walk of the MethodAnnotatedBase.
+				// This is to test that hierarchy walking of one annotation type tree will correctly mark/cache things
+				// for a different annotatin tree.
+				GetPropertiesInstance ().GetType ().GetProperties ();
+			}
+		}
+
+		[Kept]
+		class ApplyingAnnotationIntroducesTypesToApplyAnnotationToEntireType
+		{
+			[Kept]
+			[KeptMember (".ctor()")]
+			[KeptAttributeAttribute (typeof (DynamicallyAccessedMembersAttribute))]
+			[DynamicallyAccessedMembers (DynamicallyAccessedMemberTypes.All)]
+			class AnnotatedBase
+			{
+			}
+
+			[Kept]
+			[KeptBaseType (typeof (AnnotatedBase))]
+			[KeptMember (".ctor()")]
+			class Derived : AnnotatedBase
+			{
+				[Kept]
+				public void Method () { }
+
+				[Kept]
+				[KeptMember (".ctor()")]
+				[KeptInterface (typeof (INestedInterface))]
+				class Nested : INestedInterface
+				{
+					[Kept]
+					public void InterfaceMethod () { }
+				}
+
+				[Kept]
+				public AnotherAnnotatedType PublicProperty { [Kept] get => null; }
+
+				[Kept]
+				[KeptMember (".ctor()")]
+				[KeptBaseType (typeof (AnnotatedBase))]
+				class NestedAnnotatedType : AnnotatedBase
+				{
+					[Kept]
+					int _field;
+				}
+
+				[Kept]
+				int _field;
+			}
+
+			[Kept]
+			interface INestedInterface
+			{
+				[Kept]
+				void InterfaceMethod ();
+			}
+
+			[Kept]
+			[KeptMember (".ctor()")]
+			[KeptBaseType (typeof (AnnotatedBase))]
+			class AnotherAnnotatedType : AnnotatedBase
+			{
+				[Kept]
+				int _field;
+			}
+
+			[Kept]
+			static AnnotatedBase GetInstance () => new Derived ();
+
+			[Kept]
+			public static void Test ()
+			{
+				Type t = GetInstance ().GetType ();
+				t.RequiresAll ();
 			}
 		}
 	}

